libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp

```diff
@@ -0,0 +1,182 @@
+
+    // glitch handling.  rangefinder readings more than RANGEFINDER_GLITCH_ALT_CM from the last good reading
+    // are considered a glitch and glitch_count becomes non-zero
+    // glitches clear after RANGEFINDER_GLITCH_NUM_SAMPLES samples in a row.
+    // glitch_cleared_ms is set so surface tracking (or other consumers) can trigger a target reset
+    const int32_t glitch_cm = alt_cm - alt_cm_glitch_protected;
+    bool reset_terrain_offset = false;
+    if (glitch_cm >= RANGEFINDER_GLITCH_ALT_CM) {
+        glitch_count = MAX(glitch_count+1, 1);
+        set_status(Status::Glitch_Detected);
+    } else if (glitch_cm <= -RANGEFINDER_GLITCH_ALT_CM) {
+        glitch_count = MIN(glitch_count-1, -1);
+        set_status(Status::Glitch_Detected);
+    } else {
+        glitch_count = 0;
+        alt_cm_glitch_protected = alt_cm;
+    }
+    if (abs(glitch_count) >= RANGEFINDER_GLITCH_NUM_SAMPLES) {
+        // clear glitch and record time so consumers (i.e. surface tracking) can reset their target altitudes
+        glitch_count = 0;
+        alt_cm_glitch_protected = alt_cm;
+        glitch_cleared_ms = now;
+        reset_terrain_offset = true;
+    }
+
+    // filter rangefinder altitude
+    const bool timed_out = now - last_healthy_ms > RANGEFINDER_TIMEOUT_MS;
+    if (alt_healthy) {
+        if (timed_out) {
+            // reset filter if we haven't used it within the last second
+            alt_cm_filt.reset(alt_cm);
+            reset_terrain_offset = true;
+            set_status(Status::Stale_Data);
+        } else {
+            alt_cm_filt.apply(alt_cm, 0.05f);
+        }
+        last_healthy_ms = now;
+    }
+
+    // handle reset of terrain offset
+    if (reset_terrain_offset) {
+        if (rotation == ROTATION_PITCH_90) {
+            // upward facing
+            terrain_offset_cm = inertial_alt_cm + alt_cm;
+        } else {
+            // assume downward facing
+            terrain_offset_cm = inertial_alt_cm - alt_cm;
+        }
+    }
+
+    Log_Write();
+}
+
+/*
+  get inertially interpolated rangefinder height. Inertial height is
+  recorded whenever we update the rangefinder height, then we use the
+  difference between the inertial height at that time and the current
+  inertial height to give us interpolation of height from rangefinder
+ */
+bool AP_SurfaceDistance::get_rangefinder_height_interpolated_cm(int32_t& ret) const
+{
+    if (!enabled_and_healthy()) {
+        return false;
+    }
+    ret = alt_cm_filt.get();
+    ret += inertial_nav.get_position_z_up_cm() - inertial_alt_cm;
+    return true;
+}
+
+void AP_SurfaceDistance::set_status(Status state_change)
+{
+    status |= uint8_t(state_change);
+}
+
+#if HAL_LOGGING_ENABLED
+void AP_SurfaceDistance::Log_Write(void) const
+{
+    // @LoggerMessage: SURF
+    // @Vehicles: Copter
+    // @Description: Surface distance measurement
+    // @Field: TimeUS: Time since system startup
+    // @Field: I: Instance
+    // @Field: St: Status
```

Comment:
You can point the docs directly at the class 
```suggestion
    // @Field: St: Status
    // @FieldBitmaskEnum: St: Status
```
Although I suspect you will have to give it a annoyingly long name so the log parser finds the right one.

You could also build the enum for just the log rather than keep the state around all the time. EG:
https://github.com/ArduPilot/ardupilot/blob/4c95a3b6ef0db64ec21f6d6f048450cc9f03d8d1/ArduPlane/quadplane.cpp#L3615-L3639

Reply:
Thanks for the example, I did not find that when I was looking.  Nice work flow, I have replicated.